File: Tree/leetcode_98.py
# Problem Link ====>>https://leetcode.com/problems/validate-binary-search-tree/description/

class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def inorder(node, prev: int) ->bool:
            if not node:
                return True
            
            if not inorder(node.left, prev):
                return False
            
            if prev is not None and prev >=node.val:
                return False
            
            prev = node.val
            return inorder(node.right, prev)
        return inorder(root, None)


if __name__ == "__main__":
    # Test case examples
    root1 = TreeNode(5, TreeNode(1), TreeNode(4))
    root2 = TreeNode(5, TreeNode(1), TreeNode(6))
    root3 = TreeNode(2, TreeNode(1), TreeNode(3))
    solution = Solution()

    print(f"Is root1 a valid BST: {solution.isValidBST(root1)}")
    print(f"Is root2 a valid BST: {solution.isValidBST(root2)}")
    print(f"Is root3 a valid BST: {solution.isValidBST(root3)}")

File: Tree/leetcode_98.cpp
// Problem Link ====>>https://leetcode.com/problems/validate-binary-search-tree/description/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        return isPossible(root, LONG_MIN, LONG_MAX);
    }

private:
    bool isPossible(TreeNode *root, long long left, long long right)
    {
        if (root == nullptr)
        {
            return true;
        }

        if (root->val <= left || root->val >= right)
        {
            return false;
        }

        return isPossible(root->left, left, root->val) && isPossible(root->right, root->val, right);
    }
};

int main(){

    TreeNode *root = new TreeNode(5);
    root->left = new TreeNode(1);
    root->right = new TreeNode(4);

    Solution solution;
    bool is_bst = solution.isValidBST(root);

    cout << "Is the tree a valid BST: " << (is_bst ? "Yes" : "No") << endl;

    delete root;

    return 0;
}